### DISCUSSION TITLE
Mindstorms Inventor Motor jerky at slow speeds

### DISCUSSION THREAD

**Discussion by lobodpav** [pybricks/support#584] [Q&A] [answered]
Hi there!

Playing with Mindstorms Inventor motors, I found them jerky at slow speeds. Attaching videos that are worth more than a thousand words 😄

This is how the motor runs on the original Lego firmware. It also produces typical Lego high-pitch whining sounds. I was able to achieve about 72 deg/s angular speed that is smooth.

https://user-images.githubusercontent.com/1579300/148643987-8f345671-f018-46af-be27-d083b1977324.mov

With Pybricks though, the motor movement is a little jerky even at 90 deg/s angular speed. Video movement seems OK, but the sound is kinda jerky. Anyway, Mindstorms firmware can achieve much quite slower angular speeds.

https://user-images.githubusercontent.com/1579300/148644034-801ce554-da00-4080-a3e5-60b86c439c59.mov

What's worse, however, is when a motor is set to `HOLD` once it stops. What happens is the motor produces a high-pitch sound, and I can hear and feel that it is "fighting" with itself in a way the motor rotates back and forth even when no pressure is made to the wheels.

https://user-images.githubusercontent.com/1579300/148644030-1f132ed0-f891-42e8-a470-272b5c0d508e.mov

This is the code I run that HOLDs the motor.

```python
from pybricks.hubs import InventorHub
from pybricks.parameters import Button, Port, Stop
from pybricks.pupdevices import Motor

hub = InventorHub()
left_wheel = Motor(Port.A)

while True:
    pressed_buttons = hub.buttons.pressed()

    if Button.LEFT in pressed_buttons:
        left_wheel.run_time(120, 3_000, then=Stop.HOLD, wait=False)

```

**Comment by laurensvalk** [comment, marked as answer]
You might want to compare the stiffness as well. By default, the Pybricks drivers have higher control constants. This suits many applications, but it’s perhaps less ideal for wheels spinning in mid air.

When you lower the pid constants, you’d get about the same smoothness.

That said, these motors aren’t well suited for anything below “10%” speed (100 deg/s). With a bit of effort we could maybe get it smooth down to “7%”, but that gain doesn’t seem worth reducing the stiffness for.

**Reply by laurensvalk** [reply]
See also https://github.com/pybricks/support/issues/366

**Reply by lobodpav** [reply]
Thanks for the hints!

But first, I need to read some theories behind PID (never studied motors in general). Found  [nice set of articles](https://medium.com/luos/dc-motor-control-welcome-to-the-jungle-7bcd76b78c7b) to begin with.

**Reply by lobodpav** [reply]
Anyway, I really love all the `Pybricks` tools. Thanks to it, I learn so much about robotics every day. I studied electronics without having too much motivation back in the day. With the Mindstorms Inventor set and Pybricks, I regret I did not pay more attention at school 😆

**Reply by laurensvalk** [reply]
Glad to hear you're having fun :smile: 

If you want to do a simple experiment, read out the current values, and then set `kp` to half of what it was originally. Maybe cut back `kd` by a bit too.

And for this application, where there's no load, you can set `ki` to 0. That should take a way that extra start/stop when holding.

**Reply by lobodpav** [reply]
Will give it a try, thanks! But first I need to finish the reading to understand what all the Proportional. Integral, and Derivative components mean 😃
